modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteClusterSnapshotSelfTest.java

```diff
@@ -0,0 +1,734 @@
+            IgniteCheckedException.class,
+            "Snapshot creation has been finished with an error");
+
+        assertTrue("Snapshot directory must be empty: " + grid0Dir,
+            !searchDirectoryRecursively(locSnpDir.toPath(), grid0Dir).isPresent());
+
+        assertTrue("Snapshot directory must be empty: " + grid1Dir,
+            !searchDirectoryRecursively(locSnpDir.toPath(), grid1Dir).isPresent());
+
+        assertTrue("Snapshot directory must exist due to grid2 has been halted and cleanup not fully performed: " + grid2Dir,
+            searchDirectoryRecursively(locSnpDir.toPath(), grid2Dir).isPresent());
+
+        IgniteEx grid2 = startGrid(2);
+
+        assertTrue("Snapshot directory must be empty after recovery: " + grid2Dir,
+            !searchDirectoryRecursively(locSnpDir.toPath(), grid2Dir).isPresent());
+
+        awaitPartitionMapExchange();
+
+        assertTrue("Snapshot directory must be empty", grid2.snapshot().getSnapshots().isEmpty());
+
+        ignite.snapshot().createSnapshot(SNAPSHOT_NAME)
+            .get();
+
+        stopAllGrids();
+
+        IgniteEx snp = startGridsFromSnapshot(2, SNAPSHOT_NAME);
+
+        assertSnapshotCacheKeys(snp.cache(dfltCacheCfg.getName()));
+    }
+
+    /** @throws Exception If fails. */
+    @Test
+    public void testClusterSnapshotWithRebalancing() throws Exception {
+        IgniteEx ignite = startGridsWithCache(2, dfltCacheCfg, CACHE_KEYS_RANGE);
+
+        TestRecordingCommunicationSpi commSpi = TestRecordingCommunicationSpi.spi(ignite);
+        commSpi.blockMessages((node, msg) -> msg instanceof GridDhtPartitionSupplyMessage);
+
+        startGrid(2);
+
+        ignite.cluster().setBaselineTopology(ignite.cluster().topologyVersion());
+
+        commSpi.waitForBlocked();
+
+        IgniteFuture<Void> fut = ignite.snapshot().createSnapshot(SNAPSHOT_NAME);
+
+        commSpi.stopBlock(true);
+
+        fut.get();
+
+        stopAllGrids();
+
+        IgniteEx snp = startGridsFromSnapshot(3, SNAPSHOT_NAME);
+
+        awaitPartitionMapExchange();
+
+        assertSnapshotCacheKeys(snp.cache(dfltCacheCfg.getName()));
+    }
+
+    /** @throws Exception If fails. */
+    @Test
+    public void testClusterSnapshotWithExplicitPath() throws Exception {
+        File exSnpDir = U.resolveWorkDirectory(U.defaultWorkDirectory(), "ex_snapshots", true);
+
+        try {
+            IgniteEx ignite = null;
+
+            for (int i = 0; i < 2; i++) {
+                IgniteConfiguration cfg = optimize(getConfiguration(getTestIgniteInstanceName(i)));
+
+                cfg.setSnapshotPath(exSnpDir.getAbsolutePath());
+
+                ignite = startGrid(cfg);
+            }
+
+            ignite.cluster().baselineAutoAdjustEnabled(false);
+            ignite.cluster().state(ACTIVE);
+
+            for (int i = 0; i < CACHE_KEYS_RANGE; i++)
+                ignite.cache(DEFAULT_CACHE_NAME).put(i, i);
+
+            forceCheckpoint();
```

Comment:
Why do we need explicit checkpoint here?

Reply:
Fixed.